hazelcast/src/main/java/com/hazelcast/config/AbstractConfigBuilder.java

```diff
@@ -143,45 +150,83 @@ private void loadAndReplaceImportElement(Node root, Node node) throws Exception
+        Properties properties = new Properties();
+        for (Node n : childElements(node)) {
+            String value = cleanNodeName(n);
+            if ("properties".equals(value)) {
+                fillProperties(n, properties);
+            }
+        }
+        ConfigReplacer replacer = (ConfigReplacer) Class.forName(replacerClass).newInstance();
+        replacer.init(properties);
+        return replacer;
+    }
+
+    private void traverseChildrenAndReplaceVariables(Node root, ConfigReplacer replacer) {
         NamedNodeMap attributes = root.getAttributes();
         if (attributes != null) {
             for (int k = 0; k < attributes.getLength(); k++) {
                 Node attribute = attributes.item(k);
-                replaceVariables(attribute);
+                replaceVariables(attribute, replacer);
             }
         }
         if (root.getNodeValue() != null) {
-            replaceVariables(root);
+            replaceVariables(root, replacer);
         }
         final NodeList childNodes = root.getChildNodes();
         for (int k = 0; k < childNodes.getLength(); k++) {
             Node child = childNodes.item(k);
-            traverseChildrenAndReplaceVariables(child);
+            traverseChildrenAndReplaceVariables(child, replacer);
         }
     }
 
-    private void replaceVariables(Node node) {
+    private void replaceVariables(Node node, ConfigReplacer replacer) {
         String value = node.getNodeValue();
         StringBuilder sb = new StringBuilder(value);
+        String replacerPrefix = "$" + replacer.getPrefix() + "{";
         int endIndex = -1;
-        int startIndex = sb.indexOf("${");
+        int startIndex = sb.indexOf(replacerPrefix);
         while (startIndex > -1) {
             endIndex = sb.indexOf("}", startIndex);
             if (endIndex == -1) {
-                LOGGER.warning("Bad variable syntax. Could not find a closing curly bracket '}' on node: " + node.getLocalName());
+                LOGGER.warning("Bad variable syntax. Could not find a closing curly bracket '}' for prefix " + replacerPrefix
+                        + " on node: " + node.getLocalName());
                 break;
             }
 
-            String variable = sb.substring(startIndex + 2, endIndex);
-            String variableReplacement = getProperties().getProperty(variable);
+            String variable = sb.substring(startIndex + replacerPrefix.length(), endIndex);
+            String variableReplacement = replacer.getReplacement(variable);
             if (variableReplacement != null) {
                 sb.replace(startIndex, endIndex + 1, variableReplacement);
                 endIndex = startIndex + variableReplacement.length();
             } else {
-                LOGGER.warning("Could not find a value for property  '" + variable + "' on node: " + node.getLocalName());
+                LOGGER.warning("Could not find a replacement for '" + sb.substring(startIndex, endIndex + 1) + "' on node: "
```

Comment:
Can't this lead to treating encoded strings as plain texts if there is an exception during decryption (like not existing passwords file)? Shouldn't we prevent starting the member in this case?

Reply:
I usually prefer the fail fast approach, but here I'm trying to be as much backward compatible as possible in the behavior.

The current implementation allows to have more encryption replacers defined and any (or none) of them can succeed.

Reply:
After discussion with @blazember and other interested parties it was decided to introduce a new attribute which will allow to control the behavior (default: fail-fast; optional: continue processing with variable string not replaced)